### PR TITLE
New custom rule for checking patch compliance

### PR DESCRIPTION
**Pull request by lahirian**
Hi,

This is my first time submitting a pull request on this repo. Please let me know if I need to modify something. 

This is a new custom rule that I wrote to find out if a set of ec2 instances (filtered based on a given tag and value) is compliant to a SSM patch baseline.

### CODE REVIEW

File: python/ssm_patch_compliance.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        "annotation": "This resource is compliant with the rule."
+    }
+
+
+
+def lambda_handler(event, context):
+
+    print(event)
+    invoking_event = json.loads(event["invokingEvent"])
+    #configuration_item = invoking_event["configurationItem"]
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]: Why? This rule should be configuration changes triggered, as far I understand the intend.

File: python/ssm_patch_compliance.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    configuration_item = {'resourceType':'AWS::EC2::Instance'}
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]: Should not be hardcoded

File: python/ssm_patch_compliance.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import json
+import boto3
+
+
+
+
+APPLICABLE_RESOURCES = ["AWS::EC2::Instance"]
+
+
+def validate_patch_baseline(instanceid):
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]: Split this function in two which return a boolean: 1) checking the presence of an agent, 2) checking the presence of an unpatched instance.

File: python/ssm_patch_compliance.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    client = boto3.client('ssm')
+    filters = [
+		{'Key':'ComplianceType','Values':['Patch'],'Type':'EQUAL'},
+		{'Key':'Status','Values':['NON_COMPLIANT'],'Type':'EQUAL'}
+	     ]
+    response = client.list_compliance_items(Filters=filters,ResourceIds=[instanceid],ResourceTypes=['ManagedInstance'])
+    print(response)
+    if len(response['ComplianceItems']) > 0:
+        #link_to_ssm = '<a href="https://console.aws.amazon.com/systems-manager/managed-instances/' + instanceid + '/configurationCompliance>"'
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]: Remove

File: python/ssm_patch_compliance.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        security_critical = ''
+        
+        for item in response['ComplianceItems']:
+            
+            if 'Classification' in item['Details']:
+                if item['Details']['Classification'] == 'Security':
+                    #security_critical = security_critical + item['Title'] + ":"
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]: Remove

File: python/ssm_patch_compliance.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                    security_critical = " Several pending security patches."
+        violation="NON_COMPLIANT: " + instanceid  + ":" + security_critical
+        print(violation)
+        return violation
+
+    response = client.describe_instance_patch_states(InstanceIds=[instanceid])
+    
+    if len(response['InstancePatchStates']) == 0:
```
Comment (jongogogo) [inline python/ssm_patch_compliance.py]:
```
if not response['InstancePatchStates']:
```

File: RULES.md [hunk 7]
```diff
@@ -331,3 +331,11 @@ Description: Check that S3 buckets have default encryption enabled.
 
 Trigger Type: ```Change Triggered```
 Required Parameter: ```None```
+
+### 38. Check the patch compliance of a set of instances based on their tag and SSM
```
Comment (jongogogo) [inline RULES.md]: Please remove this file from your PR. We are working on getting this file auto-generated.

**Comment by jongogogo** [comment]
As been launched as managed rules: https://docs.aws.amazon.com/config/latest/developerguide/ec2-managedinstance-patch-compliance-status-check.html

**Comment by lahirian** [comment]
Thanks for the review. Sorry I just noticed this and missed replying on previous comments.

File: python/ssm_patch_compliance.py (hunk 1 above)
Comment (lahirian) [reply python/ssm_patch_compliance.py]: At the same time there may be occasions when we want to run the config rule manually to ensure that all instances are patch compliant ie in event of some major security vulnerability announcement. But I think this gives a good foundation for users to customize as needed.

**Comment by jongogogo** [comment]
No worries. Keep pushing PR :)

**Pull request closed without merging**